Tools/scripts/freeze_modules.py

```diff
@@ -113,67 +169,168 @@ def parse_frozen_spec(rawspec, knownids=None, section=None):
             pkgname = modname
             def iter_subs():
                 for frozenid, pyfile, ispkg in resolved:
-                    assert not knownids or frozenid not in knownids, (frozenid, rawspec)
+                    assert not knownids or frozenid not in knownids, (frozenid, spec)
                     if pkgname:
                         modname = frozenid.replace(pkgid, pkgname, 1)
                     else:
                         modname = frozenid
                     yield frozenid, pyfile, modname, ispkg, section
             submodules = iter_subs()
 
-    spec = (frozenid, pyfile or None, modname, ispkg, section)
-    return spec, submodules
+    info = (frozenid, pyfile or None, modname, ispkg, section)
+    return info, submodules
 
 
-def parse_frozen_specs(rawspecs=FROZEN):
-    seen = set()
-    for section, _specs in rawspecs:
-        for spec in _parse_frozen_specs(_specs, section, seen):
-            frozenid = spec[0]
-            yield spec
-            seen.add(frozenid)
+#######################################
+# frozen source files
+
+class FrozenSource(namedtuple('FrozenSource', 'id pyfile frozenfile')):
+
+    @classmethod
+    def from_raw(cls, raw, destdir=MODULES_DIR):
+        section = None
+        if isinstance(raw, cls):
+            return raw.resolve_frozen(destdir), section
+        elif isinstance(raw, FrozenSource):
+            raise NotImplementedError(raw)
```

Comment:
Why the two cases? There are no subclasses of FrozenSource, so cls is always going to be that.

Reply:
Yeah, it's unnecessary.  I've cleaned it up.